### PR TITLE
Computed metrics

### PR DESCRIPTION
**Pull request by paologallinaharbur**
Computed metrics were affected by a bug preventing them to work at all 

Fix newrelic/nri-kubernetes#311

### CODE REVIEW

File: src/kubelet/kubelet_test.go [hunk 1]
```diff
@@ -118,8 +110,6 @@ func kubeletExclusions() []exclude.Func {
 				return limitsRequestsRegex.MatchString(spec.Name)
 			},
 		),
-		// Exclude metrics that depend on limits when those limits are not set.
-		exclude.Exclude(exclude.Groups("pod", "container"), exclude.Dependent(utilizationDependencies)),
```
Comment (roobre) [inline src/kubelet/kubelet_test.go:122]: This line was intended to exclude utilization metrics only when pod does not have request/limits, should we really remove it?

File: src/metric/definition.go [hunk 2]
```diff
@@ -1017,11 +1027,46 @@ func isPersistentVolume() definition.FetchFunc {
...
+func convertValue(v interface{}) (float64, error) {
+	switch v := v.(type) {
+	case uint64:
+		return float64(v), nil
+	case int:
+		return float64(v), nil
+	case prometheus.GaugeValue:
+		return float64(v), nil
+	case float64:
+		return v, nil
```
Comment (roobre) [inline src/metric/definition.go:1059]: I'd say we can shorten the switch statement like this:
```suggestion
	case uint64, int, float64, prometheus.GaugeValue:
		return float64(v), nil
```

File: src/metric/definition.go [hunk 3]
```diff
@@ -1017,11 +1027,46 @@ func isPersistentVolume() definition.FetchFunc {
...
+	case definition.FetchedValues:
+		if len(v) == 1 {
+			for _, k := range v {
+				return convertValue(k)
+			}
+		}
+		return 0, fmt.Errorf("unable to convert FetchedValues")
```
Comment (roobre) [inline src/metric/definition.go:1066]: And get rid of one indentation level by erroring out earlier
```suggestion
		// If retrieved metric is a list of values, we can only handle the case where
		// the list has only one.
		if len(v) != 1 {
			return 0, fmt.Errorf("unable to convert FetchedValues")
		}
		for _, m := range v {
			return convertValue(m)
		}
```

**Review by roobre — COMMENTED** [review]
Should we add a test case covering the multi-values thing?

File: src/metric/definition.go (hunk 2 above)
Comment (paologallinaharbur) [reply src/metric/definition.go:1059]: Screen Shot 2022-01-07 at 12 42 29 PM
That was my first implementation, but sadly go compiler is not smart enough. I found it really weird as well that was not possible

File: src/metric/definition.go (hunk 3 above)
Comment (paologallinaharbur) [reply src/metric/definition.go:1066]: makes sense!

File: src/kubelet/kubelet_test.go (hunk 1 above)
Comment (paologallinaharbur) [reply src/kubelet/kubelet_test.go:122]: That is not needed anymore, I can bring it back in case it is needed in the future :)

Comment (paologallinaharbur) [reply src/kubelet/kubelet_test.go:122]: @roobre let me know what you think

File: src/metric/definition.go (hunk 2 above)
Comment (roobre) [reply src/metric/definition.go:1059]: fuck, you're totally right.

File: src/kubelet/kubelet_test.go (hunk 1 above)
Comment (roobre) [reply src/kubelet/kubelet_test.go:122]: I'm not sure I understand, why is it not needed anymore 🤔 

If I have a pod without `requests.limits.memory` (in k8s), what would the value of `memoryUtilization` be?

Comment (paologallinaharbur) [reply src/kubelet/kubelet_test.go:122]: that in the current scenario it is not needed. We do not need that exclude rule to pass tests with the current dataset

Comment (roobre) [reply src/kubelet/kubelet_test.go:122]: That should not be the case, we should miss this metrics as many (most) of the containers have no `resources` specified.

I think there's a bug on the not running exclusion, I'm working on it now.

Comment (roobre) [reply src/kubelet/kubelet_test.go:122]: Done, just pushed a commit to make the test fail 😈 @paologallinaharbur

**Comment by roobre** [comment]
Btw, given the excessive amount of metrics we need to exclude, I'll be happy to use `Silently()` on this one.

**Review by roobre — APPROVED** [review]

**Pull request merged**